File: joptimize/test/src/examples/narrow/Parametric2.scala
package joptimize.examples.narrow

object Parametric2 {
  def main(x: Int, y: Int): Int = {
    call(Bar, x) + call(Qux, y)
    // (x + 1) + (y + 2)
  }

  def call[T <: Foo](f: T, n: Int): Int = call2(f, n)
  def call2[T <: Foo](f: T, n: Int): Int = f.inc(n)
}

File: joptimize/test/src/examples/narrow/Parametric2.java
package joptimize.examples.narrow;

class Parametric2 {
  static int main(int x, int y) {
    return call(new Bar(), x) + call(new Qux(), y);
    // (x + 1) + (y + 2)
  }

  static <T extends Foo> int call(T f, int n){ return call2(f, n); }
  static <T extends Foo> int call2(T f, int n){ return f.inc(n); }
}
